### PR TITLE
Support logical properties in style

### PR DESCRIPTION
**Pull request by Manishearth**
Adds support for the logical block-end/inline-start/etc properties. These properties (like `border-block-end-color`) map to "physical" properties (e.g. `border-top-color`) depending on the writing mode.

Todo:

 - [x] Handle shorthands 
 - [x] Make geckolib setters work
 - [x] Handle padding/offset logical properties
 - [x] Perhaps handle `-block-size`, `-inline-size` type logical properties?
 - [x] Tests?

This will overall add 16 new longhands and 4 new shorthands, taking a big bite out of the [remaining properties work](https://manishearth.github.io/css-properties-list/?stylo=hide&servo=hide&firefox=only&chrome=show&mdn=false&alexa=false) 

f? @emilio @SimonSapin

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14120)

### CODE REVIEW

**Comment by Manishearth** [comment]
I'm unsure about the cascade behavior; currently the logical properties cascade normally, but their `set_foo` function redirects. I'm not sure if more changes need to be made here.

In particular, I'm not sure how to handle `inherit` values.

**Comment by Manishearth** [comment]
Manually tested a bit, seems to work. Some prefs need to be enabled for writing-mode to work.

**Comment by Manishearth** [comment]
I created a nice testcase for this, which shows up as

screen shot 2016-11-08 at 12 26 35 pm


(Styled version of https://drafts.csswg.org/css-writing-modes/#logical-to-physical)


<details>
```html
<!DOCTYPE html>
<html>
<head>
<style> 

#block-start div.outer div {
  border-block-start-color: red;
  border-block-start-width: 2px;
  border-block-start-style: dotted;
}
#block-end div.outer div {
  border-block-end-color: red;
  border-block-end-width: 2px;
  border-block-end-style: dotted;
}
#inline-start div.outer div {
  border-inline-start-color: red;
  border-inline-start-width: 2px;
  border-inline-start-style: dotted;
}
#inline-end div.outer div {
  border-inline-end-color: red;
  border-inline-end-width: 2px;
  border-inline-end-style: dotted;
}

#body tr td:nth-child(2) div {
  writing-mode: horizontal-tb;
}
#body tr td:nth-child(3) div {
  writing-mode: horizontal-tb;
}
#body tr td:nth-child(4) div {
  writing-mode: vertical-rl;
}
#body tr td:nth-child(5) div {
  writing-mode: vertical-rl;
}
#body tr td:nth-child(6) div {
  writing-mode: vertical-lr;
}
#body tr td:nth-child(7) div {
  writing-mode: vertical-lr;
}
/* gecko */

#body tr td:nth-child(8) div {
  writing-mode: sideways-lr;
}
#body tr td:nth-child(9) div {
  writing-mode: sideways-lr;
}

/* servo */
/*
#body tr td:nth-child(8) div {
  writing-mode: vertical-lr;
  text-orientation: sideways;
}
#body tr td:nth-child(9) div {
  writing-mode: vertical-lr;
  text-orientation: sideways;
}
*/

#body tr td:nth-child(2n+2) div {
  direction: ltr;
}
#body tr td:nth-child(2n+3) div {
  direction: rtl;
}
td {
  padding: 50px;
  background-color: yellow;
}
.top {
  border-bottom: 2px dotted green;
}
.right {
  border-left: 2px dotted green;

}

.bottom {
  border-top: 2px dotted green;

}
.left {
  border-right: 2px dotted green;

}

</style>
</head>
<body>
<table>
<thead>
  <tr><th>writing mode</th><th colspan=2>horizontal-tb</th><th colspan=2>vertical-rl</th><th colspan=2>vertical-lr</th><th colspan=2>sideways-lr</th></tr>
  <tr><th>direction</th><th>ltr</th><th>rtl</th><th>ltr</th><th>rtl</th><th>ltr</th><th>rtl</th><th>ltr</th><th>rtl</th></tr>
</thead>
<tbody id=body>
  <tr id=block-start><th>block-start</th>   <td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=left>left</div></div></td></tr>
  <tr id=block-end><th>block-end</th>       <td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=right>right</div></div></td></tr>
  <tr id=inline-start><th>inline-start</th> <td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=top>top</div></div></td></tr>
  <tr id=inline-end><th>inline-end</th>     <td><div class=outer><div class=right>right</div></div></td><td><div class=outer><div class=left>left</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td><td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=top>top</div></div></td><td><div class=outer><div class=bottom>bottom</div></div></td></tr>
</tbody>
</table>

</body>
</html>

```
</details>

The test passes when the green dotted line is opposite the red one.

Servo doesn't support `writing-mode: sideways-*` (but does support `text-orientation: sideways` and uses it in the `WritingMode` struct to determine box flow (this is wrong)), so you have to uncomment the servo decl blocks (and comment the gecko ones) for this to work in servo.

However, writing modes in general are broken in Servo layout so instead we get something like

screen shot 2016-11-08 at 12 07 37 pm

After inspecting this a bit more and looking at individual cases it seems like this PR handles the logical mapping correctly, _except_ for `inline-*` in `vertical-lr` mode. Small testcase:

<details>
```html
<!DOCTYPE html>
<html>
<head>
<style> 

#testing {
  writing-mode: vertical-lr;
  width: 300px;
  height: 300px;
  position: relative;
  top: 100px;
  left: 100px;

  background-color: yellow;
}
#testing div {

  border-inline-start-color: red;
  border-inline-start-width: 5px;
  border-inline-start-style: dotted;
  border-top: blue 5px dotted;
  width: 100px;
  height: 100px;
  background-color: green;
}
</style>
</head>
<body>
<div id=testing>
<div>AAAAA </div>
</div>

<script type="text/javascript">
  
  console.log(getComputedStyle(document.querySelector("#testing div")).borderBottomColor);
</script>


</body>
</html>
```
</details>

Should print the default color (white or currentColor) to the console, instead prints red. When displaying (hard to get it to display), the blue border should coincide with the red one, but they're on opposite sides in Servo.

Probably just needs swapping in the WritingMode methods. Though we may need more changes to support sideways-*.

**Comment by Manishearth** [comment]
Hm, when I try in geckolib the bug goes away. I think it's just servo doing crazy things with the layout.

screen shot 2016-11-08 at 12 26 35 pm

**Comment by Manishearth** [comment]
Filed https://github.com/servo/servo/issues/14144 for the writing mdoe stuff.

**Comment by Manishearth** [comment]
This PR is more or less complete.  Manually tested in stylo and works. Servo regurgitates writing modes so it's a bit harder to test, but no reason why it shouldn't work. The current test is manual; I'll see if I can get it into the CSS wg testsuite. A similar test can be written with JS and getComputedStyle, not sure if it's worth it.




r? @SimonSapin

**Comment by Manishearth** [comment]
Reftests put up in https://bugzilla.mozilla.org/show_bug.cgi?id=1316172 . The exact reftest won't work for stylo until we sort out #14144.

File: components/style/properties/data.py [hunk 1]
```diff
@@ -63,12 +72,20 @@ def maybe_cast(self, type_str):
         return "as " + type_str if self.needs_cast() else ""
 
 
+def arg_to_bool(arg):
+    if isinstance(arg, bool):
+        return arg
+    else:
+        assert arg == "True" or arg == "False"
```
Comment (emilio) [inline components/style/properties/data.py]: nit: No need for the `else` branch, you can just dedent this block. Also, `assert arg in ["True", "False"]`?

File: components/style/properties/helpers.mako.rs [hunk 2]
```diff
@@ -607,3 +611,70 @@
...
+<%def name="logical_setter_helper(name)">
+    <%
+        side = None
+        size = None
+        maybe_side = [s for s in LOGICAL_SIDES if s in name]
+        maybe_size = [s for s in LOGICAL_SIZES if s in name]
+        if len(maybe_side) == 1:
+            side = maybe_side[0]
+        elif len(maybe_size) == 1:
+            size = maybe_size[0]
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: Probably just `assert size is not None or side is not None`, and just use `else` below?

File: components/style/properties/helpers.mako.rs [hunk 3]
```diff
@@ -607,3 +611,70 @@
...
+        } else {
+            ${caller.inner(physical_ident=to_rust_ident(name.replace(size, physical_size[0])))}
+        }
+    % else:
+        <% raise Exception("Don't know what to do with logical property %s" % name) %>
+    % endif
+</%def>
+
+<%def name="logical_setter(name, need_clone=False)">
+
```
Comment (emilio) [inline components/style/properties/helpers.mako.rs]: nit: extra newline?

File: components/style/properties/longhand/position.mako.rs [hunk 4]
```diff
@@ -137,41 +138,28 @@ ${helpers.predefined_type("flex-basis",
...
-
-${helpers.predefined_type("max-height",
-                          "LengthOrPercentageOrNone",
-                          "computed::LengthOrPercentageOrNone::None",
-                          "parse_non_negative",
-                          animatable=True)}
+% for (size, logical) in ALL_SIZES:
+    // width, height, block-size, inline-size
+    ${helpers.predefined_type("%s" % size,
+                            "LengthOrPercentageOrAuto",
```
Comment (emilio) [inline components/style/properties/longhand/position.mako.rs]: nit: Alignment, here and below.

**Review by emilio — APPROVED** [review]
r=me with those nits addressed :)

**Comment by Manishearth** [comment]
Addressed, except for changing the exception to an assert

**Comment by emilio** [comment]
Ok, it's fine :)

@bors-servo r+

**Pull request merged**